### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by mzaremba04**

### CODE REVIEW

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+## How to run docker daemon in debug mode
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: What is it for? Where can I find debug info?

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+For windows there is only one [graph driver](https://blog.mobyproject.org/where-are-containerds-graph-drivers-145fc9b7255) which is called windowsfilter. It is basically wrapper for hcsshim requests which is wrapper for HCS requests.
+
+Command discussed: docker run -id --name example microsoft/windowsservercore powershell
+
+Calling POST /v1.39/containers/create?name=example
+
+Container creation is pretty straightforward:
+
+1. Function ContainerCreate from container.go is called with configuration parameters. Most relevant function in chain is create function from daemon\graphdriver\windows\windows.go
+1. Request for layer mount path of parent layers, in this case windowsservercore image, is sent to hcsshim (hcsshim::GetLayerMountPath). This wiil be just path to directoy as they are read-only.
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'is'?

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: Does "they are read-only" imply that "this wiil be just path to directoy"?

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'directoy' -> 'directory'

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+1. Resuest for creation of scratch layer (in other words read-write layer, which is for some unknown reason called sandbox layer in hscshim) is sent to hcsshim. Note that id of this layer will and should be an id of a container. (hcsshim::CreateScratchLayer)
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'Resuest' -> 'Request'

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'hscshim' -> 'hcsshim'

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "will and should be"? Maybe just "is"?

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+Calling POST /v1.39/containers/f7288bb4dc15d08dfe6cddb1830b06d88e1384967b6cda19165c5019a9575d3d/wait?condition=next-exit
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: Is this point part of container creation? Shouldn't it have its own number?

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+This just means "wait for the next time the state changes to a non-running state", which should happen immediately after creation is completed.
+
+This is it for creation part, now request for starting container is sent. During the starting phase if networking is provided it will also be created.
+Calling POST /v1.39/containers/f7288bb4dc15d08dfe6cddb1830b06d88e1384967b6cda19165c5019a9575d3d/start
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: What does "f7288bb4dc15d08dfe6cddb1830b06d88e1384967b6cda19165c5019a9575d3d" mean?

File: doc/For developers/Technical documentation/Docker_HNS.md (hunk 4 above)
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: What does "f7288bb4dc15d08dfe6cddb1830b06d88e1384967b6cda19165c5019a9575d3d" mean?

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+1. At the beginning container base filesystem is mounted. This is done by calling Get() function from windowsfilter driver on newly created RW layer. hcsshim::ActivateLayer is the call that mounts RW layer (wiil appear as a volume on the host) and hscshim::PrepareLayer intializes the filesystem filter for that container.
+
+Next part is network configuration:
+ 
+## What exactly happens during attaching container to network
+
+### Compartments
+
+Compartments are Windows network namespaces. To list comnpartments run Get-NetCompartment. there is one default compartment for OS which has CompartmentId = 1 and type Native. Container compartments will have CompartmentType Unspecified. To link compartment to a container look into CompartemntDesciption field, it will have format "\Container_(container_id)"
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'comnpartments' -> 'compartments'

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: 'there' -> 'There'

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "CompartemntDesciption" -> "CompartmentDescription".

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+### Interafaces
+
+To list container interfaces run  get-netipinterface -addressfamily Ipv4 -includeallcompartments | Where InterfaceAlias -Like "*Container NIC*"
+To get interace from specific compartment check execute get-netipinterface -addressfamily Ipv4 -includeallcompartments -compartmentId id.
+
+## Fun Facts
+
+Docker on Windows doesn't support restoring containers when daemon dies
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: Not really fun.

Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: You may add information about Docker version that these remarks refer to.

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+### Compartments
+
+Compartments are Windows network namespaces. To list comnpartments run Get-NetCompartment. there is one default compartment for OS which has CompartmentId = 1 and type Native. Container compartments will have CompartmentType Unspecified. To link compartment to a container look into CompartemntDesciption field, it will have format "\Container_(container_id)"
+
+### Interafaces
+
+To list container interfaces run  get-netipinterface -addressfamily Ipv4 -includeallcompartments | Where InterfaceAlias -Like "*Container NIC*"
+To get interace from specific compartment check execute get-netipinterface -addressfamily Ipv4 -includeallcompartments -compartmentId id.
+
+## Fun Facts
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md:61]: Please use punctuation consistently.

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+Docker on Windows doesn't support restoring containers when daemon dies
+
+To see the id of image layer run docker image inspect imagename and look under GraphDriver key
+
+To see layers just list C:\ProgramData\docker\windowsfilter\ directory.
+
+To see layerchain for container see C:\ProgramData\docker\windowsfilter\continer_id\layerchain.json
+
+Winddows doesn't behave well when mounted RW layer id isnt't the same as id of container
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "Winddows" -> "Windows"

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
+## How to run docker daemon in debug mode
+
+Just run dockerd.exe -D in console or add "debug": true to config and restart daemon.
+
+## What exactly happens during container creation
+
+For windows there is only one [graph driver](https://blog.mobyproject.org/where-are-containerds-graph-drivers-145fc9b7255) which is called windowsfilter. It is basically wrapper for hcsshim requests which is wrapper for HCS requests.
+
+Command discussed: docker run -id --name example microsoft/windowsservercore powershell
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: What do you think about making structure of this section a bit more clear? E.g.:

```
Command "docker run -id --name example microsoft/windowsservercore powershell" results in sending the following POST request "/v1.39/containers/create?name=example".

It triggers the following actions:
I. Creating container
1.
2. ...
II. Starting container
1.
2. ...
```

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+## What exactly happens during attaching container to network
+
+### Compartments
+
+Compartments are Windows network namespaces. To list comnpartments run Get-NetCompartment. there is one default compartment for OS which has CompartmentId = 1 and type Native. Container compartments will have CompartmentType Unspecified. To link compartment to a container look into CompartemntDesciption field, it will have format "\Container_(container_id)"
+
+### Interafaces
+
+To list container interfaces run  get-netipinterface -addressfamily Ipv4 -includeallcompartments | Where InterfaceAlias -Like "*Container NIC*"
+To get interace from specific compartment check execute get-netipinterface -addressfamily Ipv4 -includeallcompartments -compartmentId id.
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "check execute" - what do you mean? Probably "check" is not necessary here.

File: doc/For developers/Technical documentation/Docker_HNS.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+Next part is network configuration:
+ 
+## What exactly happens during attaching container to network
+
+### Compartments
+
+Compartments are Windows network namespaces. To list comnpartments run Get-NetCompartment. there is one default compartment for OS which has CompartmentId = 1 and type Native. Container compartments will have CompartmentType Unspecified. To link compartment to a container look into CompartemntDesciption field, it will have format "\Container_(container_id)"
+
+### Interafaces
```
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "Interafaces" -> "Interfaces"

File: doc/For developers/Technical documentation/Docker_HNS.md (hunk 9 above)
Comment (jacek-izykowski) [inline doc/For developers/Technical documentation/Docker_HNS.md]: "doesn't behave well" - what does it mean? What is happening in such a case? What actions can result in those ids being different?

**Review by sodar — APPROVED** [review]

**Pull request merged**